### DISCUSSION TITLE
How would I use a remote persistant storage?

### DISCUSSION THREAD

**Discussion by benjick** [pmndrs/zustand#742] [Q&A]
Hello!

I wanna use persist to sync data remotely so I can continue working on another device. How would I solve this?

I created this example storage:

```ts
import { StateStorage } from "zustand/middleware";

const url = "https://example.com/api/state/";

export const remoteStorage: StateStorage = {
  getItem: async (name) => {
    const res = await fetch(url + userId);
    const json = await res.json();
    return json;
  },
  setItem: async (name, value) => {
    await fetch(url + userId, {
      method: "POST",
      body: JSON.stringify({ value }),
    });
  },
  removeItem: async (name) => {
    await fetch(url + userId, {
      method: "DELETE",
    });
  },
};
```

but I'm not sure how I would get `userId` into the actual storage. Any suggestions?

**Comment by dai-shi** [comment]
Interesting use case. Not sure if it's fully supported. What I can think of is encoding `userId` in `name`.
Cc: @AnatoleLucet

**Reply by benjick** [reply]
I thought about that as well, but I'm not sure how I would change `name` during runtime

**Reply by dai-shi** [reply]
`setOptions` could do. But, it's not designed for this use case as you might guess.

https://github.com/pmndrs/zustand/blob/1dce3d2f140374630c4a64a897858e9308379c37/src/middleware/persist.ts#L311-L315

**Comment by AnatoleLucet** [comment]
~~Not 100% sure, but since your storage is async, you should be able to use `myStore.getState().userId` inside your storage methods~~ Nop, definitely not possible

**Reply by benjick** [reply]
That makes sense, I will try it out!

**Reply by benjick** [reply]
It works fine for `setItem` but for `getItem` I sadly get the following error:

> TypeError: _state__WEBPACK_IMPORTED_MODULE_2__.useStore is undefined

**Reply by benjick** [reply]
Okay I saw your update now. Too bad! Thanks for taking the time

**Comment by benjick** [comment]
So what I ended up doing for now is basically just `const userId = localStorage.getItem('userId');`
